Sming/Arch/Esp8266/Components/libc/include/sys/pgmspace.h

```diff
@@ -0,0 +1,174 @@
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2015 by Skurydin Alexey
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * pgmspace.h - Support for reading flash memory (for old GCC versions)
+ *
+ ****/
+
+#pragma once
+
+#include <sys/features.h>
+
+#ifdef __NEWLIB__
+
+#include_next <sys/pgmspace.h>
+
+#else
+
+#include <c_types.h>
+#include <esp_attr.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup pgmspace Flash support functions
+ * @ingroup flash
+ * @{
+ */
+
+/**
+ * @brief Identifies a char pointer as living in flash memory
+ * Use to clarify code
+ */
+#define PGM_P const char*
+
+/**
+ * @brief Remove ?
+ */
+#define PRIPSTR "%s"
+
+/**
+ * @name AVR-compatible pgmspace types
+ * Deprecated in esp8266-arduino https://github.com/esp8266/Arduino/pull/4619/
+ * @{
+ */
+typedef void prog_void;
+typedef char prog_char;
+typedef unsigned char prog_uchar;
+typedef int8_t prog_int8_t;
+typedef uint8_t prog_uint8_t;
+typedef int16_t prog_int16_t;
+typedef uint16_t prog_uint16_t;
+typedef int32_t prog_int32_t;
+typedef uint32_t prog_uint32_t;
+/** @} */
+
+#define PROGMEM __attribute__((aligned(4))) ICACHE_FLASH_ATTR
+
+// flash memory must be read using 32 bit aligned addresses else a processor exception will be triggered
+// order within the 32 bit values are
+// --------------
+// b3, b2, b1, b0
+//     w1,     w0
+
+#define pgm_read_with_offset(addr, res)                                                                                \
```

Comment:
There is a new version of this macro in the newlib folder which uses 1 fewer assembly instruction.

Reply:
Thanks, I'll need to review that module in some detail I think.

I note that you've used `asm` in those macros, but that won't compile with std=c++X, requires `__asm__` instead - I think because it uses extended syntax. (I've just added a #define asm __asm__ but it's a bit of a hack...)

Whilst I'm at it, our PROGMEM implementation includes `aligned((4))` but your implementation does not. Is there a specific reason for that?

Reply:
That's very odd.  The PROGMEM macro builds and runs fine w/ gcc=c++14 in the main Arduino repo.  Never looked into it, really, since it just worked.

We don't align PROGMEM, but we do align PSTRs and I've rewritten several of the _P string functions to use this to speed things up in newlib:
`  #define PSTR(s)            (__extension__({static const char __c[] __attribute__((__aligned__(4))) PROGMEM = (s); &__c[0];}))`

Also, related to PSTRs, recently we adjusted them to allow the linker to combine duplicated strings and save lots of memory in certain apps:  
https://github.com/esp8266/Arduino/pull/6565

I see now that *that* change did not filter back into newlib, darn it.  I need to do a PR there...

Reply:
The asm thing works because Arduino builds with `-std=gnu++11` but we're using `-std=c++11` and `-std=c++17` to try and keep with the standard.

The de-duplication is interesting, will definitely check that out thanks. My thought on the PROGMEM alignment was simply that if there's no justification for every using it unaligned then it does help to simplify things a little. My preference is to try to steer clear of 'raw' progmem stuff where possible. See #1974 for example, using counted flash strings with C++ objects. Plus there's https://github.com/Chris--A/PGMWrap. Lots of ways to leverage C++ and avoid the low-level stuff!

Reply:
@slaff I'm inclined to leave our sys/pgmspace.h as-is for this PR as it affects builds with existing (old) toolchains. The de-duplication can probably be applied in otger areas as well.